command/replace.go

```diff
@@ -0,0 +1,170 @@
+// This is free and unencumbered software released into the public
+// domain.  For more information, see <http://unlicense.org> or the
+// accompanying UNLICENSE file.
+
+package command
+
+import (
+	"fmt"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/nelsam/gxui"
+	"github.com/nelsam/gxui/themes/basic"
+	"github.com/nelsam/vidar/commander/bind"
+	"github.com/nelsam/vidar/commander/input"
+)
+
+type Replace struct {
+	driver gxui.Driver
+	theme  *basic.Theme
+
+	find      *findBox
+	replace   *findBox
+	status    gxui.Label
+	editor    SelectionEditor
+	applier   Applier
+	edits     []input.Edit
+	is_select bool
+
+	input <-chan gxui.Focusable
+}
+
+func NewReplace(driver gxui.Driver, theme *basic.Theme) *Replace {
+	replacer := &Replace{}
+	replacer.Init(driver, theme)
+	return replacer
+}
+
+func (f *Replace) Init(driver gxui.Driver, theme *basic.Theme) {
+	f.driver = driver
+	f.theme = theme
+
+}
+
+func (f *Replace) Start(control gxui.Control) gxui.Control {
+	f.editor = findEditor(control)
+	if f.editor == nil {
+		return nil
+	}
+	f.is_select = false
+
+	f.find = newFindBox(f.driver, f.theme)
+	f.find.OnTextChanged(func([]gxui.TextBoxEdit) {
+		f.editor.Controller().ClearSelections()
+
+		f.is_select = true
+		needle := f.find.Text()
+		if len(needle) == 0 {
+			f.status.SetText("Start typing to search")
+			return
+		}
+		haystack := f.editor.Text()
+		start := 0
+		var selections gxui.TextSelectionList
+
+		count := utf8.RuneCountInString(needle)
+		length := len(needle)
+		pos := 0
+		for next := strings.Index(haystack, needle); next != -1; next = strings.Index(haystack[start:], needle) {
+			pos += utf8.RuneCountInString(haystack[start : start+next])
+			selection := gxui.CreateTextSelection(pos, pos+count, false)
+			selections = append(selections, selection)
+			pos += count
+			start += (next + length)
+		}
+		f.editor.Select(selections)
+		f.status.SetText(fmt.Sprintf("%s: %d results found", needle, len(selections)))
+	})
+	f.find.OnKeyPress(func(ev gxui.KeyboardEvent) {
+		switch ev.Key {
+		case gxui.KeyEscape:
+			if f.is_select {
+				f.editor.Controller().ClearSelections()
+			}
+		}
+	})
```

Comment:
I guess it's also will be appropriate for find commands.

Reply:
Probably.  I am pretty sure that I want both find and replace to work differently in the future, but there's work to do before then.  GXUI doesn't really know about highlights without selections, right now, and it needs to before we can cycle through find results.

We could probably implement a `find-query-replace` with the existing functionality, but it would be pretty poor.  I think we just need to add to GXUI before we make those changes I want.